transaction-view/src/message_header_meta.rs

```diff
@@ -0,0 +1,123 @@
+use {
+    crate::{
+        bytes::read_byte,
+        result::{Result, TransactionParsingError},
+    },
+    solana_sdk::message::MESSAGE_VERSION_PREFIX,
+};
+
+/// A byte that represents the version of the transaction.
+#[derive(Copy, Clone, Default)]
+#[repr(u8)]
+pub enum TransactionVersion {
+    #[default]
+    Legacy = u8::MAX,
+    V0 = 0,
+}
+
+/// Meta data for accessing message header fields in a transaction view.
+pub(crate) struct MessageHeaderMeta {
+    /// The offset to the first byte of the message in the transaction packet.
+    pub(crate) offset: u16,
+    /// The version of the transaction.
+    pub(crate) version: TransactionVersion,
+    /// The number of signatures required for this message to be considered
+    /// valid.
+    pub(crate) num_required_signatures: u8,
+    /// The last `num_readonly_signed_accounts` of the signed keys are
+    /// read-only.
+    pub(crate) num_readonly_signed_accounts: u8,
+    /// The last `num_readonly_unsigned_accounts` of the unsigned keys are
+    /// read-only accounts.
+    pub(crate) num_readonly_unsigned_accounts: u8,
+}
+
+impl MessageHeaderMeta {
+    pub fn try_new(bytes: &[u8], offset: &mut usize) -> Result<Self> {
+        // Get the message offset.
+        // We know the offset does not exceed packet length, and our packet
+        // length is less than u16::MAX, so we can safely cast to u16.
+        let message_offset = *offset as u16;
+
+        // Read the message prefix byte if present. This byte is present in V0
+        // transactions but not in legacy transactions.
+        // The message header begins immediately after the message prefix byte
+        // if present.
+        let message_prefix = read_byte(bytes, offset)?;
+        let (version, message_header_offset) = if message_prefix & MESSAGE_VERSION_PREFIX != 0 {
+            let version = message_prefix & !MESSAGE_VERSION_PREFIX;
+            match version {
+                // add one byte for prefix byte. Do not need to check for
+                // overflow here, because we have already checked.
+                0 => (TransactionVersion::V0, message_offset.wrapping_add(1)),
```

Comment:
You could have just done `*offset += 1` here, instead of creating a new variable.

Reply:
Yeah I don't think we need to have new variable here. We've already advanced the byte, we actually need to "go back" in the case of legacy instead of v1 since we read the byte here:

```rust
        let message_prefix = read_byte(bytes, offset)?;
```

Reply:
let me know what you think of this: 2b30a847714087f5ebec31b3939d5775c188d0aa


Instead of adjusting the offset I just adjusted how we read the number of required signatures.

Legacy txs it will just be the byte we tried reading for the message prefix.

Reply:
that should actually save us from re-reading the byte; though i'm not sure that would significantly affect peformance.